orttraining/orttraining/core/session/training_session.cc

```diff
@@ -1124,5 +1144,660 @@ std::unordered_set<std::string> TrainingSession::GetTrainableModelInitializers(
   return trainable_initializers;
 }
 
+TrainingSession::~TrainingSession() {
+//#if defined(USE_NCCL) && defined(USE_NCCL_P2P)
+//  auto& nccl_service = cuda::NcclService::GetInstance();
+//  nccl_service.Terminate();
+//#endif
+}
+
+#if defined(USE_NCCL) && defined(USE_NCCL_P2P)
+// Create NCCL's communication plan. In runtime, we will provide details such
+// as pointer to sent/recieved data and the size of the data in byte. See how
+// Send and Recv call SubmitSendAndWait and SubmitRecvAndWait, respectively.
+void PipelineTrainingSession::LaunchNcclService(const int pipeline_stage_id) {
+  ORT_ENFORCE(pipeline_stage_id >= 0, "Pipeline stage ID cannot be negative.");
+  auto& nccl_service = cuda::NcclService::GetInstance();
+
+  // Create NCCL communication plan. The plan is a vector of communication task group.
+  // Each communication task group contains tasks which should be done in parallel.
+  for (auto& slot : pipeline_schedule_.GetSchedule(pipeline_stage_id)) {
+    if (!slot.HasCommute()) {
+      continue;
+    }
+    // Create communication tasks done in parallel.
+    nccl_service.PlanNewGroupStart();
+    for (auto& task : slot.GetTasks()) {
+      if (task.type == pipeline::PipelineTask::Type::Send) {
+        // In this time slot, stage "pipeline_stage_id" sendss data to "task.peer_rank".
+        nccl_service.PlanSend(task.peer_rank);
+      } else if (task.type == pipeline::PipelineTask::Type::Recv) {
+        // In this time slot, stage "pipeline_stage_id" recieves data from "task.peer_rank".
+        nccl_service.PlanRecv(task.peer_rank);
+      }
+    }
+    // Mark the end of a parallel communication task group.
+    nccl_service.PlanNewGroupEnd();
+  }
+  // Mark the end of the entire communication plan.
+  nccl_service.PlanEnd();
+
+  // Launch NCCL service to execute the plan.
+  nccl_service.Launch();
+}
+#endif
+
+Status PipelineTrainingSession::InsertPipelineOpsAndCreateFakeOutputs(
+    const std::unordered_set<std::string>& initializer_names_to_preserve) {
+  ORT_RETURN_IF_ERROR(TransformGraphForPipeline(
+      true,
+      initializer_names_to_preserve,
+      pipeline_context_.sliced_schema,
+      pipeline_context_.expected_output_names,
+      model_->MainGraph(),
+      pipeline_context_.pipeline_tensor_names
+      ));
+  return DoPostLoadProcessing(*model_);
+}
+
+Status PipelineTrainingSession::ConfigureForTraining(
```

Comment:
>PipelineTrainingSession::ConfigureForTraining [](start = 7, length = 45)

seems this function shares a lot of functions with its base class, and most of it are not pipeline related. Maybe consider refactor it to re-use the existing code?

Reply:
Rewrite ConfigureForTraining. Pipeline logics are in the following two functions now.
```
  Status PartitionGraphForPipeline(
      const int32_t pipeline_stage_id,
      const optional<TrainingConfiguration::PipelineConfiguration>& pipeline_config,
      const optional<TrainingConfiguration::DistributedConfiguration>& distributed_config,
      const std::unordered_set<std::string>& weight_names_to_train,
      std::unordered_set<std::string>& filtered_config_weight_names_to_train) override;

  Status SetEventSynchronization(
      const int32_t pipeline_stage_id,
      const optional<TrainingConfiguration::PipelineConfiguration>& pipeline_config,
      const optional<TrainingConfiguration::DistributedConfiguration>& distributed_config,
      const std::unordered_set<std::string>& weight_names_to_train,
      optional<TrainingConfigurationResult::PipelineConfigurationResult>& pipeline_config_result) override;
```

---
In reply to: [538815253](https://github.com/microsoft/onnxruntime/pull/5815#discussion_r538815253) [](ancestors = 538815253)